exonum-java-binding/core/rust/src/storage/entry.rs

```diff
@@ -44,10 +44,14 @@ pub extern "system" fn Java_com_exonum_binding_core_storage_indices_EntryIndexPr
         let name = utils::convert_to_string(&env, name)?;
         Ok(handle::to_handle(
             match handle::cast_handle::<View>(view_handle).get() {
-                ViewRef::Snapshot(snapshot) => {
-                    IndexType::SnapshotIndex(Index::new(name, &*snapshot))
-                }
-                ViewRef::Fork(fork) => IndexType::ForkIndex(Index::new(name, fork)),
+                ViewRef::Snapshot(snapshot) => IndexType::SnapshotIndex(
+                    Index::from_access(snapshot, name.into())
+                        .expect("Index type does not match specified one"),
```

Comment:
Can we use #new that doesn't return a Result but Self? 

Is it the only possible Error ("Index type does not match specified one" — will a user see the actual in the description?)? Is this error documented in the core?

Reply:
#new is private, so no. There is a bunch of methods like `fn get_proof_list<I, V>(self, addr: I) -> ProofListIndex<Self::Base, V>`, but they all use the same code as here (with `unwrap` instead of `expect`).

Error message contains also index name, expected and actual type. The only reason I added `expect` - to pass existing Java tests in BaseIndexProxyIntegrationTests.

> Is it the only possible Error?

There is also second variant called `Custom`, but I can't find if it's used anywhere in Core.

Reply:
@slowli Are any other errors possible in this method?

Reply:
But will that error message be included in the exception message that we throw in case of panic here, or disappear? If unwrap provides that, it'd be better. 

Or, better, can we just return Error and `unwrap_exc_or` will throw exception with Error#to_string?

While we are here, please document in the `unwrap_exc_or` documentation what nested Results mean, because each time this method appears one has to dig in to understand that.

Reply:
Current exception message is: `Index type does not match specified one: AccessError { addr: IndexAddress { name: "test_index", bytes: None }, kind: WrongIndexType { expected: List, actual: ValueSet } }`

Returning error is... complicated, because `unwrap_exc_or` is not working with any kinds of errors, but rather with `jni` errors.

Reply:
Do I understand correctly, that if we replace expect(msg) with unwrap, it will be
"AccessError { addr: IndexAddress { name: "test_index", bytes: None }, kind: WrongIndexType { expected: List, actual: ValueSet } }"?

Reply:
If that's correct, IMO it's good enough in this rare case.

Reply:
Yep, it initially was. I will change it.